src/node/timeoffsets.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <logging.h>
+#include <node/interface_ui.h>
+#include <node/timeoffsets.h>
+#include <sync.h>
+#include <util/translation.h>
+#include <warnings.h>
+
+#include <algorithm>
+#include <chrono>
+#include <deque>
+
+using namespace std::chrono_literals;
+
+void TimeOffsets::Add(std::chrono::seconds offset)
+{
+    LOCK(m_mutex);
+
+    if (m_offsets.size() >= N) {
+        m_offsets.pop_front();
+    }
+    m_offsets.push_back(offset);
+}
+
+std::chrono::seconds TimeOffsets::Median() const
+{
+    LOCK(m_mutex);
+
+    // Only calculate the median if we have 5 or more offsets
+    if (m_offsets.size() < 5) return 0s;
+
+    auto sorted_copy = m_offsets;
+    std::sort(sorted_copy.begin(), sorted_copy.end());
+    return sorted_copy[m_offsets.size() / 2];  // approximate median is good enough, keep it simple
+}
+
+bool TimeOffsets::WarnIfOutOfSync() const
+{
+    // when median == std::numeric_limits<int64_t>::min(), calling std::chrono::abs is UB
+    auto median{std::max(Median(), std::chrono::seconds(std::numeric_limits<int64_t>::min() + 1))};
+    if (std::chrono::abs(median) <= m_warn_threshold) return false;
+
+    bilingual_str msg{_("Your computer's date and time appear out of sync with the network, "
+                        "this may lead to consensus failure. Please ensure it is correct.")};
+    LogWarning("%s\n", msg.original);
```

Comment:
Grepping in my node's log for how frequent outbound connections are made - it would take about 1.5 days to make 200 new outbound connections. Since this is about the median, I guess 101 new connections after fixing the clock will be enough for us to stop the warning. That is like 20ish hours. Quite long. Can we do better?

Currently we only save the difference between our time and the peer's time at the start of the connection and we have no clue that this difference has changed due to our clock being adjusted/corrected. If we save the value of the system_clock and steady_clock at the start of the connection, in addition to the peer's offset, at a later time we can deduce that our clock has been adjusted and add that diff/adjustment to the peer's offset to immediately figure out that "now our clock is in sync with that peer's clock".

So, instead of `peer_offset`, it would be a matter of using the following instead:
`peer_offset + steady_clock_diff - system_clock_diff`, assuming:
`steady_clock_diff = steady_clock_now - steady_clock_when_connected`
`system_clock_diff = system_clock_now - system_clock_when_connected`.

Example:
---
Lets say our clock is wrong, 15 minutes ahead. When we connect to a peer at `17:00:00` our time they send us a time that is 15 minutes ago and we save:
`peer_offset = -15min`,
`system_clock_when_connected = 17:00:00`,
`steady_clock_when_connected = 1000` (some obscure value [seconds]).
One hour later, our clock has been fixed/corrected, by moving it 15 min backwards (or stopping to move for 15 min). So,
`system_clock_now = 17:45:00`,
`steady_clock_now = 4600` (the obscure value 1000 plus 3600 because 1 hour has passed).
And then:
`steady_clock_diff = 4600 - 1000 = 60 min`
`system_clock_diff = 17:45:00 - 17:00:00 = 45 min`
The difference between those is 15 minutes and adding that to `peer_offset` gives us offset 0, meaning that now our clock is in sync with the peer's clock.

Reply:
Using both system and steady clock is an interesting idea and one I hadn't considered, thanks. I'm going to look into this a bit more and will follow up shortly.

> it would take about 1.5 days to make 200 new outbound connections

FYI this PR reduces 200 -> 50 outbounds so your calculations are a bit off.